ROX-9565: support building, pushing and testing operator on prow

porridge opened the pull request:
## Description

Please look at individual commits' descriptions.

These changes should not change anything in the CircleCI-based operator tests.
At the same time, this is the complete set of changes (on the `stackrox/stackrox` repo side) for running operator CI on OpenShift CI.

The [release branch is here](https://github.com/porridge/release/pull/new/ROX-9565-operator-on-prow) and a PR will be created once this part is merged.

## Checklist
- [ ] Investigated and inspected CI test results
- [x] ~Unit test and regression tests added~
- [x] ~Evaluated and added CHANGELOG entry if required~
- [x] ~Determined and documented upgrade steps~
- [x] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Roughly the same changes (module some debug output and comments) were tested in https://github.com/openshift/release/pull/30630

SimonBaeumer commented on scripts/ci/jobs/openshift_4_operator_e2e_tests.py:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env -S python3 -u
+
+"""
+Run operator e2e tests in a openshift 4 cluster provided via a hive cluster_claim.
```
Do you documentation on this? Honestly I don't understand it.

porridge replied:
I'm just trying to be consistent with the another script in the same directory from which I cargo-culted.

SimonBaeumer commented on scripts/ci/lib.sh:
```diff
@@ -202,6 +202,103 @@ push_main_image_set() {
+                v=""
+            fi
+            oc image mirror "${operator_image_srcs[$idx]}" "${registry}/${image}:${v}${tag}"
+            (( idx++ )) || true
```
Why is this function the only one which has an index in its loop?

mtesseract replied:
Looks like we need to process `operator_image_set` and `operator_image_srcs` in parallel. So we use the `idx` for indexing in the second array.

SimonBaeumer commented on scripts/ci/lib.sh:
```diff
@@ -202,6 +202,103 @@ push_main_image_set() {
+    }
+
+    if [[ "$brand" == "STACKROX_BRANDING" ]]; then
+        local destination_registries=("quay.io/stackrox-io")
+    elif [[ "$brand" == "RHACS_BRANDING" ]]; then
+        local destination_registries=("quay.io/rhacs-eng")
```
Does it make sense to have the registry repos configurable by env variables?

mtesseract replied:
Well, we need credentials for pushing and these are passed in via the CI vault and they are, of course, for a fixed registry. So, in any case, changing the registry would not be simply a matter of changing some registry env variable.

What I attempted the other day was also putting the registry right next to username & password into the vault, so they form a group, like:
```
IMAGE_PUSH_REGISTRY_NAME
IMAGE_PUSH_REGISTRY_USERNAME
IMAGE_PUSH_REGISTRY_PASSWORD
```

But the downside of this is that whenever you would like to output someothing like "Pushing $IMG to $REGISTRY", the registry will be redacted by OSCI. So, this is also not ideal. I would say, under these circumstances it's not an issue to hardcode the registry.

These are just my 2c.

porridge replied:
What I [alluded (probably not strongly enough) to in the commit description](https://github.com/stackrox/stackrox/pull/2471/commits/5ae1bc4a5ef1a102df992e03b5246bbd8412be43) is that this whole function is an almost exact copy of `push_main_image_set`, so I'm just being consistent with what the other does.

Unfortunately due to bash limitations **and** the inconsistent versioning scheme of our images, it would be a bit arcane to try and reuse code between them.

We can think about improving both of them, but I'd consider this to be out of scope for this PR.

SimonBaeumer commented on scripts/ci/lib.sh:
```diff
@@ -202,6 +202,103 @@ push_main_image_set() {
+        die "$brand is not a supported brand"
+    fi
+
+    local tag
+    tag="$(make --quiet -C operator tag)"
```
Why are two lines needed here?

mtesseract replied:
To not silently swallow any failures. Shellcheck also warns about that.

mtesseract replied:
for example:
```
➜  stackrox git:(porridge/ROX-9565-operator-on-prow) ✗ bash -c 'set -e; export foo=$(false); echo "error swallowed"'
error swallowed
➜  stackrox git:(porridge/ROX-9565-operator-on-prow) ✗ bash -c 'set -e; foo=$(false); export foo; echo "error swallowed"'
➜  stackrox git:(porridge/ROX-9565-operator-on-prow) ✗ 
```

Same for `local` instead of `export`.

gavin-stackrox replied:
thats a `shellcheck` suggestion: https://www.shellcheck.net/wiki/SC2155

porridge replied:
Again, I just cargo culted, thanks @mtesseract for the explanation, TIL.

SimonBaeumer commented on .openshift-ci/build/build-operator-artifacts.sh:
```diff
@@ -0,0 +1,45 @@
+
+build_operator_bundle_and_binary() {
+    # avoid a -dirty tag
+    info "Reset to remove Dockerfile modification by OpenShift CI"
```
How does the CI adjust the dockerfiles? 🤔 
This seems odd to me tbh

mtesseract replied:
ci-operator has this mechanism of injecting images into the `FROM ...` directives. Seems to be common practice in OSCI world, we also use it in asc-fleet-manager. But, it is also news to me that this works by patching the unpacked source code and leaving it like that.

gavin-stackrox replied:
CI updates the `FROM`(s)

SimonBaeumer reviewed (approved):
LGTM, good job @porridge !
Some minor comments, imho you can also address them in another PR to get this one merged.

mtesseract commented on operator/hack/build-index-image.sh:
```diff
@@ -2,9 +2,21 @@
+# Normally the same as OPM_VERSION. Here we use the upstream version for serving the index,
+# since the render-dir hack only affects the build. This helps us avoid the hassle of hosting the OPM image.
+OPM_IMAGE_VERSION="1.21.0"
+OPM_ORG="porridge"
```
This is a dirty hack, because upstream opm didn't work for us.

gavin-stackrox commented on .openshift-ci/dispatch.sh:
```diff
@@ -27,6 +27,18 @@ case "$ci_job" in
         openshift_ci_e2e_mods
         ;;
+    openshift-*-operator-e2e-tests)
+        registry_ro_login "quay.io/rhacs-eng"
```
can all this go into a `operator_e2e_test_setup()` (or something) in lib.sh?

mtesseract commented on operator/Makefile:
```diff
@@ -67,7 +70,11 @@ SCORECARD_STORAGE_IMAGE ?= $(SCORECARD_STORAGE_IMAGE_BASE):$(SCORECARD_STORAGE_I
+# TODO(ROX-11889):   Error: error running tests error creating ConfigMap namespaces "ci-op-05qx07r3" not found
+# TODO(ROX-11889): These seem to happen despite unsetting the $NAMESPACE environment variable.
+# TODO(ROX-11889): Figure out where scorecard gets the namespace name from, elliminate this source, and remove the flag here.
+SCORECARD_ARGS ?= --storage-image="$(SCORECARD_STORAGE_IMAGE)" --wait-time="$(SCORECARD_WAIT_TIME_DURATION)" --verbose --namespace default
```
Shouldn't this be `--storage-image`?

gavin-stackrox commented on scripts/ci/jobs/push-images.sh:
```diff
@@ -44,6 +44,10 @@ push_images() {
         push_context="merge-to-master"
     fi
 
+    # TODO(ROX-11889): make this unconditional once the openshift/release side is ready
```
nit: shouldn't operator be after main

porridge replied:
I explain the reason for this order in the commit description.

>  I assume that this passed in the rehearsal PR somewhere and a link to a log would be useful.

As mentioned in the PR description:
> Roughly the same changes (module some debug output and comments) were tested in https://github.com/openshift/release/pull/30630

I avoided doing this for this PR, because it would cause the operator images to be built both by prow and by circleci and clash. OTOH if I made a change here to remove the CircleCI operator jobs in this PR, then we wouldn't know if this PR didn't break them. But I guess that's OK since they are scheduled for disabling today.

gavin-stackrox reviewed (changes requested):
Looks good. Some minor nits. I assume that this passed in the rehearsal PR somewhere and a link to a log would be useful.

mtesseract commented on operator/Makefile:
```diff
@@ -67,7 +70,11 @@ SCORECARD_STORAGE_IMAGE ?= $(SCORECARD_STORAGE_IMAGE_BASE):$(SCORECARD_STORAGE_I
+# TODO(ROX-11889): The `--namespace default` is a workaround for failures such as the following when running in a pod:
+# TODO(ROX-11889):   Error: error running tests error creating ConfigMap namespaces "ci-op-05qx07r3" not found
+# TODO(ROX-11889): These seem to happen despite unsetting the $NAMESPACE environment variable.
+# TODO(ROX-11889): Figure out where scorecard gets the namespace name from, elliminate this source, and remove the flag here.
```
Could it be that it is set in the default context in the kubeconfig? As in

```
- context:
  [...]
    namespace: ...
  [...]
```

porridge replied:
The kubeconfig used by scorecard is the kubeconfig for the leased cluster, which cannot possibly mention the CI namespace. I suspect it's trying to be overly clever by looking up the namespace from some kubernetes downward API.

mtesseract commented on scripts/ci/lib.sh:
```diff
@@ -202,6 +202,103 @@ push_main_image_set() {
+
+    local operator_image_set=("stackrox-operator" "stackrox-operator-bundle" "stackrox-operator-index")
+    if is_OPENSHIFT_CI; then
+        local operator_image_srcs=("$OPERATOR_IMAGE" "$OPERATOR_BUNDLE_IMAGE" "$OPERATOR_BUNDLE_INDEX_MAGE")
```
Should this also be initialized somehow in the non-OSCI case?

mtesseract reviewed (approved):
Looking good.

gavin-stackrox commented on scripts/ci/lib.sh:
```diff
@@ -202,6 +202,103 @@ push_main_image_set() {
+                v=""
+            fi
+            oc image mirror "${operator_image_srcs[$idx]}" "${registry}/${image}:${v}${tag}"
+            (( idx++ )) || true
```
> Why the `|| true`?

I think thats from `set -e`. 
```
$ (idx=0; (( idx++ )); echo incremented)
incremented
$ (set -e; idx=0; (( idx++ )); echo incremented)
$
```

porridge commented on operator/hack/build-index-image.sh:
```diff
@@ -91,7 +103,7 @@ function fetch_opm() {
 
   OPM="${BASE_DIR}/bin/opm-${OPM_VERSION}"
-  "${SCRIPT_DIR}/get-github-release.sh" --to "${OPM}" --from "https://github.com/operator-framework/operator-registry/releases/download/v${OPM_VERSION}/${os_name}-${arch}-opm"
+  "${SCRIPT_DIR}/get-github-release.sh" --to "${OPM}" --from "https://github.com/${OPM_ORG}/operator-registry/releases/download/v${OPM_VERSION}/${os_name}-${arch}-opm"
```
> Does this take a fixed script from your personal github account?

It takes a binary built from the fork.

> WDYTH we should fork the repositories into the stack rocks organization instead of our personal accs?

In general I agree, though there is precedent in `go.mod` for personal orgs.
As mentioned in the TODO above, this is temporary.

porridge commented on operator/hack/build-index-image.sh:
```diff
@@ -2,9 +2,21 @@
 
 set -eou pipefail
+# TODO(ROX-11889): Remove the -x once we gain some confidence (after 3.72 release?)
+set -x
```
I double-checked there are none [in the log](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/30630/rehearse-30630-pull-ci-stackrox-stackrox-porridge-ROX-9569-operator-prow-upgrade-images/1549746175630905344/artifacts/build-logs/operator-artifacts.log).

porridge commented:
Created a [temporary rehearsal PR](https://github.com/openshift/release/pull/30674).
Still not opening a PR for the `openshift/release` [branch](https://github.com/porridge/release/pull/new/ROX-9565-operator-on-prow) yet because it would fail anyway until this PR is merged.

porridge commented:
@gavin-stackrox I think I addressed all your comments and [ran a rehearsal](https://github.com/openshift/release/pull/30674) (which revealed I was missing one more `export`), PTAL
